src/classes/CDL_CascadeDeleteLookups.cls

```diff
@@ -0,0 +1,349 @@
+    * @description Contains Cascade Undelete logic. The Trigger Handlers use this class to retrieve 
+    * the children of the undeleted records, and undelete the children records.
+    ********************************************************************************************************/
+    public class CascadeUndelete {
+        private Set<Id> undeletedRecordIds;
+        private CascadeUndeleteLoader loader;
+
+        /*********************************************************************************************************
+        * @description The Constructor of CascadeUndelete.
+        */
+        public CascadeUndelete() {
+            undeletedRecordIds = new Set<Id>();
+        }
+
+        /*******************************************************************************************************
+        * @description Sets the records (from the Triggers context) that are undeleted.
+        * @param List<SObject> The undeleted records (from the Triggers context).
+        * @return void
+        ********************************************************************************************************/
+        public void setUndeletedRecords(List<SObject> undeletedRecords) {
+            this.undeletedRecordIds = new Map<Id, SObject>(undeletedRecords).keySet();
+        }
+
+        /*******************************************************************************************************
+        * @description Sets the object that loads the children records.
+        * @param CascadeUndeleteLoader The Cascade Undelete loader
+        * @return void.
+        ********************************************************************************************************/
+        public void setLoader(CascadeUndeleteLoader loader) {
+            this.loader = loader;
+        }
+
+        /*******************************************************************************************************
+        * @description Undeletes the children records.
+        * @return void.
+        ********************************************************************************************************/
+        public void undeleteRecords() {
+            CascadeUndeleteLoader currentChildrenLoader = loader;
+
+            // The Undeletion runs in a loop because some children records need to be undeleted first.
+            // Undelete all the records at once can break the references.
+            while (currentChildrenLoader != null) {
+                List<SObject> children = currentChildrenLoader.load(undeletedRecordIds);
+                if (children.isEmpty()) {
+                    break;
+                }
+
+                List<Database.UndeleteResult> undeleteResults = Database.undelete(children, false);
+                ERR_Handler.Errors errorResult = ERR_Handler.getErrors(undeleteResults, children);
+                if (errorResult.errorsExist) {
+                    ERR_Handler.processErrors(errorResult.errorRecords, 'TDTM');
```

Comment:
Does `processErrors` throw an error and stops undelete from processing any further?
If not, does `currentChildrenLoader = currentChildrenLoader.next();` selects next for only successfully undeleted children?

Reply:
@RadSFDeveloper you have a good point. "Database.undelete(children, false)" is silent, so if there is an error the current logic will go to the next group. That means, the undeletion of a parent record will fail but the logic will proceed and undelete its child. The issue is, the reference from the parent to child will be broken. 
One way I think we can solve this is if there is an error "errorResult.errorsExist()", the logic should not go to the next group.
@njjc any thoughts?

Reply:
Can we only pass the record IDs that undeleted successfully to the next round of deletion?

Reply:
Great idea @njjc. I'll implement the fix.